File: design/swift/HousePour/HPStepper.swift
// House Pour kit — HPStepper (upstream `.stepper`; kept in the kit, unused by tgsocial v1).

import SwiftUI

public struct HPStepper: View {
    @Binding var value: Int
    let range: ClosedRange<Int>
    let label: String

    public init(value: Binding<Int>, in range: ClosedRange<Int>, label: String) {
        _value = value; self.range = range; self.label = label
    }

    public var body: some View {
        HStack(spacing: HPTokens.Space.rowGap) {
            stepButton("\u{2212}", enabled: value > range.lowerBound) { value -= 1 }
            Text(String(value)).hpStyle(HPType.totals)
                .frame(minWidth: HPTokens.Space.touchMin)
            stepButton("+", enabled: value < range.upperBound) { value += 1 }
        }
        .accessibilityElement(children: .contain)
        .accessibilityLabel(label)
        .accessibilityValue(String(value))
    }

    private func stepButton(_ glyph: String, enabled: Bool, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            Text(glyph)
                .hpStyle(HPType.totals)
                .frame(width: HPTokens.Space.touchMin, height: HPTokens.Space.touchMin)
                .background(Circle().fill(HPTokens.Colors.panel))
                .overlay(Circle().strokeBorder(HPTokens.Colors.line2, lineWidth: HPTokens.borderWidth))
                .contentShape(Circle())
        }
        .buttonStyle(HPPressStyle())
        .disabled(!enabled)
        .opacity(enabled ? 1 : HPAlpha.disabled)
        .accessibilityLabel(glyph == "+" ? "Increase" : "Decrease")
    }
}

File: design/kotlin/housepour/HPStepper.kt
package ca.lucianlabs.housepour

import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp

/** Upstream stepper: two `touchMin` circles with a serif figure between. Not used in tgsocial v1; kept in the kit. */
@Composable
fun HPStepper(value: Int, onChange: (Int) -> Unit, modifier: Modifier = Modifier, min: Int = 0, max: Int = Int.MAX_VALUE) {
    Row(modifier = modifier, verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(HPTokens.Space.rowGap)) {
        StepperCircle("−", "Decrease", enabled = value > min) { onChange(value - 1) }
        HPText(value.toString(), HPTokens.Type.totalsAmount, HPTokens.Colors.ink, maxLines = 1)
        StepperCircle("+", "Increase", enabled = value < max) { onChange(value + 1) }
    }
}

@Composable
private fun StepperCircle(glyph: String, label: String, enabled: Boolean, onClick: () -> Unit) {
    Box(
        modifier = Modifier
            .size(HPTokens.Space.touchMin)
            .border(HPTokens.BORDER_WIDTH.dp, HPTokens.Colors.line2, CircleShape)
            .clickable(interactionSource = remember { MutableInteractionSource() }, indication = null, enabled = enabled, role = Role.Button, onClick = onClick)
            .semantics { contentDescription = label },
        contentAlignment = Alignment.Center,
    ) {
        HPText(glyph, HPTokens.Type.bodyStrong, if (enabled) HPTokens.Colors.ink else HPTokens.Colors.faint, maxLines = 1)
    }
}
